### DISCUSSION TITLE
logkit using logrus strange behaviour

### DISCUSSION THREAD

**Discussion by SofiaTa** [go-kit/kit#1149] [Q&A] [answered]
Hi,

1. We want to use the go-kit log [`logrus adapter`](https://github.com/go-kit/kit/blob/master/log/logrus/logrus_logger.go) library and we found something a bit strange,
every time you want to use the log API you must provide the key message "msg" for the main message otherwise it will be printed as empty

2. why the log every-time printed as `info` even though filtering correctly according

`level.Debug(logger).Log("message", "in debug mode", "key2", "foo")`

This prints: 
`
{"key2":"foo", "fields.level":{},"level":"info","message":"in debug mode","msg":""}
`

Thanks,

**Comment by ChrisHines** [comment, marked as answer]
Thanks for the questions, @SofiaTa.

For most projects the go-kit log package should be an alternative to logrus rather than a wrapper around it. As a result, I don't think the logrus adapter has been used very much and it likely has many short comings. It appears you have found two of them.

The current implementation of the logrus adapter is very simple. You can see its history in #759 and #793.

I will try to answer your questions below.

>1. We want to use the go-kit log logrus adapter library and we found something a bit strange,
every time you want to use the log API you must provide the key message "msg" for the main message otherwise it will be printed as empty

Yes, i can reproduce the behavior you describe.

This happens because the logrus data model has a special [`Message`](https://github.com/sirupsen/logrus/blob/master/entry.go#L68) field and the [JSON formatter always includes it](https://github.com/sirupsen/logrus/blob/master/json_formatter.go#L89), even if it is empty. That is the way logrus is designed and implemented.

The go-kit log package, on the other hand, does not define any special keys, allowing applications to have full control of the keys in their logs. That is a fundamental design difference between the two packages and it leaves the logrus adapter with a question of how to handle that difference. The current implementation takes the simplest approach of [not doing anything special](https://github.com/go-kit/kit/blob/master/log/logrus/logrus_logger.go#L33) to populate the logrus `Message` data.

I am not sure we should change that, but I am open to suggestions.

>2. why the log every-time printed as info even though filtering correctly according

There was no specific discussion of how to translate log levels from the go-kit log/levels package into logrus levels. Perhaps that was not a requirement of the original authors of the logrus adapter. The current implementation makes no attempt to consider the log/levels package and just always uses the logrus `Info` method. I would be happy to review a PR that improves the logrus adapter to handle log levels.

**Comment by SofiaTa** [comment]
Thanks for your quick answer @ChrisHines  👍 
From my POV I think it will be helpful if we be able to omit somehow the mandatory `msg` field which is not required for some simple cases .
we can remove the the usage of the logrus and use directly the go-kit log which could be better fit 
however when testing only this provide the similar behavior when using only go-kit vanilla logs:

```
import (
	"os"

	kitlog "github.com/go-kit/kit/log"
)

func main() {
	logger := kitlog.NewJSONLogger(kitlog.NewSyncWriter(os.Stdout))

	logger = kitlog.With(logger, "ts", kitlog.DefaultTimestampUTC, "caller", kitlog.DefaultCaller)

	logger.Log("hello")

}
```

This prints 

```
{"caller":"main.go:14","hello":"(MISSING)","ts":"2019-10-22T12:05:29.821345Z"}
```

You can see the "missing" entry in the log if you don't provide the key "msg" , I'm not really understand why its mandatory ?  there is a cases when you want to log only one statement/message , is there a way to avoid it ?

**Comment by peterbourgon** [comment]
Go kit's package log implements a concept called structured logging, which mandates logging key-value pairs rather than simple strings. Concretely, that means every call to Log requires _pairs_ of arguments, never single or odd-numbered arguments. This was an early design decision and not one that's going to change.
